examples/rl/ppo_cartpole.py

```diff
@@ -0,0 +1,267 @@
+        path_slice = slice(self.tr_start_idx, self.ptr)
+        rews = np.append(self.rew_buf[path_slice], last_val)
+        vals = np.append(self.val_buf[path_slice], last_val)
+
+        deltas = rews[:-1] + self.gamma * vals[1:] - vals[:-1]
+
+        self.adv_buf[path_slice] = discount_cumsum(deltas, self.gamma * self.lam)
+        self.ret_buf[path_slice] = discount_cumsum(rews, self.gamma)[:-1]
+
+        self.tr_start_idx = self.ptr
+
+    def get(self):
+        # Get all data of the buffer and normalize the advantages
+        self.ptr, self.tr_start_idx = 0, 0
+        adv_mean, adv_std = np.mean(self.adv_buf), np.std(self.adv_buf)
+        self.adv_buf = (self.adv_buf - adv_mean) / adv_std
+        return self.obs_buf, self.act_buf, self.adv_buf, self.ret_buf, self.logp_buf
+
+
+def mlp(x, sizes, activation=tf.tanh, output_activation=None):
+    # Build a feedforward neural network
+    for size in sizes[:-1]:
+        x = layers.Dense(units=size, activation=activation)(x)
+    return layers.Dense(units=sizes[-1], activation=output_activation)(x)
+
+
+def logprobs(logits, a):
+    # Compute the log-probabilities of taking actions a by using the logits (i.e. the output of the actor)
+    logp_all = tf.nn.log_softmax(logits)
+    logp = tf.reduce_sum(tf.one_hot(a, n_acts) * logp_all, axis=1)
+    return logp
+
+
+"""
+## Hyperparameters
+"""
+
+# Hyperparameters of the PPO algorithm
+steps_per_epoch = 4000
+epochs = 30
+gamma = 0.99
+clip_ratio = 0.2
+pi_lr = 3e-4
+vf_lr = 1e-3
+train_pi_iters = 80
+train_v_iters = 80
+lam = 0.97
+taget_kl = 0.01
+hidden_sizes = (64, 64)
+
+# True if you want to render the environment
+render = False
+
+"""
+## Initializations
+"""
+
+# Initialize the environment and get the dimensionality of the
+# observation space and the number of possible actions
+env = gym.make("CartPole-v0")
+obs_dim = env.observation_space.shape[0]
+n_acts = env.action_space.n
+
+# Initialize the buffer
+buf = Buffer(obs_dim, steps_per_epoch)
+
+# Initialize the actor and the critic as keras models
+obs_in = keras.Input(shape=(obs_dim,), dtype=tf.float32)
+logits = mlp(obs_in, list(hidden_sizes) + [n_acts], tf.tanh, None)
+actor = keras.Model(inputs=obs_in, outputs=logits)
+v = tf.squeeze(mlp(obs_in, list(hidden_sizes) + [1], tf.tanh, None), axis=1)
+critic = keras.Model(inputs=obs_in, outputs=v)
+
+# Initialize the policy and the value function optimizers
+pi_optimizer = keras.optimizers.Adam(learning_rate=pi_lr)
+v_optimizer = keras.optimizers.Adam(learning_rate=vf_lr)
+
+# Initialize the observation, episode return and episode length
+obs, ep_ret, ep_len = env.reset(), 0, 0
+
+"""
+## Train
+"""
+# Iterate over the number of epochs
+for epoch in range(epochs):
+    # Initialize the sum of the returns, lengths and number of episodes for each epoch
+    sum_ret = 0
+    num_ep = 0
+
+    # Iterate over the steps of each epoch
+    for t in range(steps_per_epoch):

```

Comment:
Move as much inner code as possible to one or several standalone functions, then decorate them with `@tf.function` compilation decorator. You may also try `@tf.function(jit_compile=True)` to see if it speeds things up.

Reply:
Hello Francois, thanks for the useful comments. I changed the variable names and I also moved the code for updating the policy and the value function on standalone functions using `@tf.function.`  The `jit_compile=True` argument did not speed things up, however.